File: typescript/src/n0102_binary_tree_level_order_traversal.ts
import { Queue } from '@datastructures-js/queue';

import type { TreeNode } from './utils/tree';

export function levelOrder(root: TreeNode | null): number[][] {
  if (!root) {
    return [];
  }
  const result = [];
  const queue = new Queue([root]);
  while (!queue.isEmpty()) {
    const level = [];
    const len = queue.size();
    for (let i = 0; i < len; i++) {
      const node = queue.pop();
      level.push(node.value);
      if (node.left) {
        queue.push(node.left);
      }
      if (node.right) {
        queue.push(node.right);
      }
    }
    result.push(level);
  }
  return result;
}

File: rust/src/n0102_binary_tree_level_order_traversal.rs
use std::{cell::RefCell, collections::VecDeque, rc::Rc};

use crate::utils::tree::TreeNode;

pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    if root.is_none() {
        return vec![];
    }
    let mut result = vec![];
    let mut queue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::from([root.unwrap()]);
    while !queue.is_empty() {
        let len = queue.len();
        let mut level = Vec::with_capacity(len);
        for _ in 0..len {
            let node = queue.pop_front().unwrap();
            level.push(node.borrow().value);
            if let Some(l) = node.borrow().left.clone() {
                queue.push_back(l);
            }
            if let Some(r) = node.borrow().right.clone() {
                queue.push_back(r);
            }
        }
        result.push(level);
    }
    result
}
